### DISCUSSION TITLE
The match ends, but the results are not recorded...

### DISCUSSION THREAD

**Discussion by Bunker56** [PhlexPlexico/G5API#220] [Help]
G5API-out-0.log:
::ffff:172.20.0.2 - - [20/Aug/2022:22:10:03 +0000] "GET /mapstats/12 HTTP/1.1" 404 29 "https://mmr.bunker56.ru/match/12" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.81 Safari/537.36"
[0mPUT /match/12/map/0/demo/upload/NB3SL6X4AX9GVU4KB59QQBBZ [33m404[0m 9.510 ms - 46[0m
::ffff:172.20.0.2 - - [20/Aug/2022:22:10:11 +0000] "PUT /match/12/map/0/demo/upload/NB3SL6X4AX9GVU4KB59QQBBZ HTTP/1.1" 404 46 "-" "sm-ripext/1.3.1"
[0mGET /vetoes/12 [33m404[0m 2.023 ms - 30[0m
::ffff:172.20.0.2 - - [20/Aug/2022:22:10:13 +0000] "GET /vetoes/12 HTTP/1.1" 404 30 "https://mmr.bunker56.ru/match/12" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.81 Safari/537.36"

Console log:
Completed GOTV demo "12_map1_de_breach.dem", recording time 802.9
StopRecording: 2687 frames flushed in 0.59 seconds
L 08/21/2022 - 03:10:11: [G5WS.smx] [ERR] Demo request failed, HTTP status code: 404

[07] REST in Pawn (1.3.1): Provides HTTP and JSON natives for plugins
21 "G5WS - Get5 Web Stats" (3.1.2) by phlexplexico
22 "Get5" (0.10.0-dev) by splewis, nickdnk & PhlexPlexico

What could be the reason?  My mistake?

**Comment by PhlexPlexico** [comment]
Please don't self-compile the get5 builds. Please use the latest nightlies instead as they intend to be more stable and don't have WIP features.

As for the match not ending, I don't see any logs for from the API even saying it was being called. Please set `get5_debug 32` and try and replicate this again and send the game server logs.

**Reply by Bunker56** [reply]
Nightly 0.10.0-2a5b3f9 - This?
OK. I'l try...

**Reply by PhlexPlexico** [reply]
Yes. It should call match ending after a certain delay. Per the docs on get5 make sure you're not adjusting any `tv_` settings in your configs.http://splewis.github.io/get5/dev/gotv/

**Comment by Bunker56** [comment]
Completed GOTV demo "14_map1_de_dust2.dem", recording time 346.2
StopRecording: 2687 frames flushed in 0.60 seconds
L 08/22/2022 - 04:15:47: [get5] Calling Get5_OnDemoFinished()
L 08/22/2022 - 04:15:47: [G5WS] About to enter UploadDemo. SO YES WE ARE. Our match ID is 14
L 08/22/2022 - 04:15:47: [G5WS] Our URL is: http://mmr.bunker56.ru/api/match/%s/map/%d/demo
L 08/22/2022 - 04:15:47: [G5WS] Trying to create request to url http://mmr.bunker56.ru/api/match/14/map/0/demo
L 08/22/2022 - 04:15:47: [G5WS] Our api url: http://mmr.bunker56.ru/api/
L 08/22/2022 - 04:15:47: [G5WS] Our demo string: 14_map1_de_dust2.dem
L 08/22/2022 - 04:15:47: [G5WS] Our URL is: http://mmr.bunker56.ru/api/match/%s/map/%d/demo/upload/%s
L 08/22/2022 - 04:15:47: [G5WS] Trying to create request to url http://mmr.bunker56.ru/api/match/14/map/0/demo/upload/6BCY14WIKJ6XEKNCZOMZ52K8
L 08/22/2022 - 04:15:47: [G5WS] Uploading demo to server...
L 08/22/2022 - 04:15:47: [G5WS] COMPLETE!
L 08/22/2022 - 04:15:47: [get5] Calling Get5_OnEvent(data={
    "matchid": "14",
    "event": "demo_finished",
    "map_number": 0,
    "filename": "14_map1_de_dust2.dem"
})
L 08/22/2022 - 04:15:47: [G5WS.smx] [ERR] Demo request failed, HTTP status code: 404

configurations checked.. Installed your ready-made plugins - the same result.... Pull your images - the same result.... Backups are copied normally.
???
"Demo request failed" - where does the request go?

**Reply by PhlexPlexico** [reply]
Hmm odd. Would you mind trying version 4.0.0 in the PR that's there? You may have to compile it yourself. But that should fix any issues. I think because it's a 404 it's just not finding the route.

**Reply by Bunker56** [reply]
> it's just not finding the route.

This is just understandable, but where to look for it?

And about 4.0.0, I will definitely try it today, it's very interesting...

**Reply by PhlexPlexico** [reply]
Well, the only reason it will 404 is if the map stats object is missing, which means the map stats weren't created on match init. 

There was a bug I found when converting everything from rest in pawn to steamworks again, so that very well may be fixed in 4.0.0.

**Reply by Bunker56** [reply]
I will try it tonight, I will inform you about the result

**Reply by Bunker56** [reply]
Completed GOTV demo "19_map1_de_dust2.dem", recording time 418.1
StopRecording: 2687 frames flushed in 0.57 seconds
L 08/23/2022 - 01:58:13: [get5] Calling Get5_OnDemoFinished()
L 08/23/2022 - 01:58:13: [G5WS] About to enter UploadDemo. SO YES WE ARE. Our match ID is 19
L 08/23/2022 - 01:58:13: [G5WS] Trying to create request to url http://mmr.bunker56.ru/api/match/19/map/0/demo
L 08/23/2022 - 01:58:13: [G5WS] Added param key=60QV2W0YQAS2U15CLTW8646B to request
L 08/23/2022 - 01:58:13: [G5WS] Our api url: http://mmr.bunker56.ru/api/
L 08/23/2022 - 01:58:13: [G5WS] Our demo string: 19_map1_de_dust2.dem
L 08/23/2022 - 01:58:13: [G5WS] Added param demoFile=19_map1_de_dust2.dem to request
L 08/23/2022 - 01:58:13: [G5WS] Trying to create request to url http://mmr.bunker56.ru/api/match/19/map/0/demo/upload/60QV2W0YQAS2U15CLTW8646B
L 08/23/2022 - 01:58:13: [G5WS] Uploading demo to server...
L 08/23/2022 - 01:58:13: [get5] Calling Get5_OnEvent(data={
    "matchid": "19",
    "event": "demo_finished",
    "map_number": 0,
    "filename": "19_map1_de_dust2.dem"
})
L 08/23/2022 - 01:58:14: [SM] Exception reported: Handle 7af020c is invalid (error 3)
L 08/23/2022 - 01:58:14: [SM] Blaming: G5WS.smx
L 08/23/2022 - 01:58:14: [SM] Call stack trace:
L 08/23/2022 - 01:58:14: [SM]   [0] CloseHandle
L 08/23/2022 - 01:58:14: [SM]   [1] Line 242, G5WS.sp::DemoCallback

**Reply by Bunker56** [reply]
And this can't happen because I work through a proxy via HTTP, and it doesn't work out any other way. If I try to do everything via HTTPS on a proxy, I get HTTP ERROR 503

**Reply by Bunker56** [reply]
If you just try to somehow disable automatic certificate receipt in Caddy...

**Reply by PhlexPlexico** [reply]
503 means there's an error on the server, not because of redirects. If you get that, it's still contacting the API. If you want to, there are ways to disable HTTPS certificate generation that has already been described [here](https://github.com/PhlexPlexico/G5API/issues/218#issuecomment-1209839461). The handle closing has been fixed. This is only the demo uploading that's failing, so it shouldn't be an issue if you don't need it at least. But I just pushed out one more change to the plugin that should fix it.

**Reply by Bunker56** [reply]
All this is very odd, because everything worked fine with get5 (0.9.0) until the latest Valve updates.

**Reply by Bunker56** [reply]
Isn't that critical?  [G5WS.smx] Failed to create logo directory: materials/panorama/images/tournaments/teams

**Reply by PhlexPlexico** [reply]
No, it just means it can't create the logo directory. If you're on Linux you need to adjust the folder permissions (+rw to the user running the server). 

By the sounds of it, if the match is ending and the scores are being reported (and the map stats exist), then everything is working as intended. Make sure you're on the latest docker images if using Docker as well. A CS update would not break any of this :)

**Comment by Bunker56** [comment]
I'm sorry for the persistence, but I don't understand what's going on. 
All images are recent, at the output in G5API-error-0.log I get this message:
TypeError: cb is not a function
    at /Get5API/node_modules/passport/lib/sessionmanager.js:85:16
    at /Get5API/node_modules/express-session/session/store.js:54:5
    at /Get5API/node_modules/@redis/client/dist/lib/client/index.js:381:13
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
The message appears when I log out of my account

It doesn't get to the recording of the veto, maybe it's all about the Docker version (I have Docker version 20.10.17, build 100c701), docker-compose.yml yours
CS:GO has just been installed.
[SM] Displaying 12 extensions:
[01] Automatic Updater (1.10.0.6545): Updates SourceMod gamedata files
[02] Webternet (1.10.0.6545): Extension for interacting with URLs
[03] CS Tools (1.10.0.6545): CS extended functionality
[04] BinTools (1.10.0.6545): Low-level C/C++ Calling API
[05] SDK Tools (1.10.0.6545): Source SDK Tools
[06] Top Menus (1.10.0.6545): Creates sorted nested menus
[07] Client Preferences (1.10.0.6545): Saves client preference settings
[08] SQLite (1.10.0.6545): SQLite Driver
[09] SteamWorks Extension (1.2.3): Exposes SteamWorks functions to Developers
[10] Regex (1.10.0.6545): Provides regex natives for plugins
[11] SDK Hooks (1.10.0.6545): Source SDK Hooks
[12] REST in Pawn (1.3.1): Provides HTTP and JSON natives for plugins
[SM] Listing 21 plugins:
  01 "Admin Menu" (1.10.0.6545) by AlliedModders LLC
  02 "Basic Comm Control" (1.10.0.6545) by AlliedModders LLC
  03 "Client Preferences" (1.10.0.6545) by AlliedModders LLC
  04 "Basic Votes" (1.10.0.6545) by AlliedModders LLC
  05 "Fun Votes" (1.10.0.6545) by AlliedModders LLC
  06 "Anti-Flood" (1.10.0.6545) by AlliedModders LLC
  07 "Basic Commands" (1.10.0.6545) by AlliedModders LLC
  08 "Reserved Slots" (1.10.0.6545) by AlliedModders LLC
  09 "Fun Commands" (1.10.0.6545) by AlliedModders LLC
  10 "Sound Commands" (1.10.0.6545) by AlliedModders LLC
  11 "Basic Ban Commands" (1.10.0.6545) by AlliedModders LLC
  12 "Basic Info Triggers" (1.10.0.6545) by AlliedModders LLC
  13 "Player Commands" (1.10.0.6545) by AlliedModders LLC
  14 "Basic Chat" (1.10.0.6545) by AlliedModders LLC
  15 "Admin File Reader" (1.10.0.6545) by AlliedModders LLC
  16 "Admin Help" (1.10.0.6545) by AlliedModders LLC
  17 "MapChooser" (1.10.0.6545) by AlliedModders LLC
  18 "Map Nominations" (1.10.0.6545) by AlliedModders LLC
  19 "Rock The Vote" (1.10.0.6545) by AlliedModders LLC
  20 "Get5" (0.10.0-8142e4c) by splewis, nickdnk & PhlexPlexico
  21 "G5WS - Get5 Web Stats" (3.1.2) by phlexplexico
Version 4.0.0 - the same result. 
I've already broken my whole head, I can't get to the truth, what could be the matter? Tell me where to dig?

**Reply by PhlexPlexico** [reply]
That error message has nothing to do with the game server. You can see in the error it mentions passport, which is to do with logging into the website.

What you need to do, so I can help you, is to create a match, set `get5_debug 32` in your game server console, and then reproduce the error. Once that's done, on your game server go to `addons/sourcemod/logs` and grab get5_debug.log and look at the errors, and send *the entire file* here.

**Reply by Bunker56** [reply]
OK. Tonight.

**Reply by Bunker56** [reply]
If the result is the same, I will try to downgrade the Docker version, because everything worked fine, it all started after reinstalling Linux (Debian 11.4). if anything, I will inform you additionally.

**Comment by Bunker56** [comment]
Hi! I tested everything again, the result is that get 5 works without errors, G5WS seems to work too, but most likely the PUT request disappears somewhere, because there is no veto entry in the database. Look at logs. And I can't understand yet where he specifically disappears... How do I track it?
[csgoserver-console.log](https://github.com/PhlexPlexico/G5API/files/9440439/csgoserver-console.log)
[get5_debug.log](https://github.com/PhlexPlexico/G5API/files/9440440/get5_debug.log)
[G5API-out-0.log](https://github.com/PhlexPlexico/G5API/files/9440441/G5API-out-0.log)
[errors_20220829.log](https://github.com/PhlexPlexico/G5API/files/9440442/errors_20220829.log)
[G5API-error-0.log](https://github.com/PhlexPlexico/G5API/files/9440443/G5API-error-0.log)

Maybe all the problems are in RIP, since I have to work via HTTP, and in the past there have already been problems because of this...

**Reply by PhlexPlexico** [reply]
First off, make sure you update your get5 configs. There's things crashing that are unrelated (such as demo uploads). .10 of get5 is going to be out shortly.

Secondly, use version 4.0.0 of G5WS, as it no longer has a need for RIP, it uses SteamWorks as well.

Looking at your front-end, it appears that you have some modifications with it as well, so I'm not really able to help if you've made any changes to the API as well. Are you able to use the public panel to test this and see if it works? Rule out any issues of the API from another host, and see if it's a networking issue.

The vetoUpdate route is a POST command, and it appears that 

version 4.0.0 has this correct as well.

Here, this has a bit more useful debug statements for G5WS when commands are created. Rest In Pawn is not needed, only SteamWorks and it should work as intended. If there's an issues not reaching routes, it could be that the API your using is modified or out of date. Please try with the public panel (either https://phlex.avidgamers.me or https://get5.phlexplexi.co) and let me know if that works.

Here's the plugin [G5WS.zip](https://github.com/PhlexPlexico/G5API/files/9440478/G5WS.zip)

**Reply by Bunker56** [reply]
It looks like it's cutting me somewhere, because I use your images, Docker just reinstalled, I didn't compile anything myself... Everything is tip-top on your panel

**Comment by Bunker56** [comment]
Everything is in perfect order! I figured it out - the problem was with certificates, we changed hosting and there was a certificate attached to the domain, and my server also received a certificate, that's where the discrepancy was... Double certificate... :)

**Reply by Bunker56** [reply]
By the way, are all of the last get 5 working normally?

**Reply by PhlexPlexico** [reply]
The latest get5 nightly should be the release candidate for 0.10. if you use that, you'll be up to date once it gets released. Lots of new features and bug fixes in it, so I would recommend you start using that :) and glad to hear you got it figured out!!

**Reply by Bunker56** [reply]
Very big thanks for help and for you сreation!..

**Comment by Bunker56** [comment]
Plugin glitch ?
![BO5](https://user-images.githubusercontent.com/91856368/187103621-68dba29c-05c8-4f7a-a96e-fea45bc59a6d.JPG)

BO1 and BO3 - normally...

**Reply by PhlexPlexico** [reply]
Appears to be a bug with the plugin, should be fixed soon.

**Reply by PhlexPlexico** [reply]
https://github.com/splewis/get5/suites/8022064266/artifacts/344451661 try this

**Reply by PhlexPlexico** [reply]
https://github.com/splewis/get5/releases/tag/v0.10.0-3d8e91c er this one

**Comment by Bunker56** [comment]
I apologize for the offtop, but I want to ask a knowledgeable person where I can find what the 'contribution score' is and how it is calculated. I dug up the entire Internet, but no one can give a clear answer...

**Reply by PhlexPlexico** [reply]
The contribution score is the final score in the scoreboard per player, so using this match as an example https://phlex.avidgamers.me/match/3864 each kill counts as 2 points in the scoreboard, so the one user has 6 points. That's all it is. You can just search "contribution" to see what all the default values are here https://totalcsgo.com/commands

**Reply by Bunker56** [reply]
Thanks!

**Comment by Bunker56** [comment]
Hi! I would really like to compile a fresh G5WS 4.0.0 , but I get this error:
G5WS.sp(222) : error 017: undefined symbol "IsClientCoaching"
G5WS.sp(750) : warning 203: symbol is never used: "GetClientCoachingSide"

Where can there be a discrepancy? get 5 compiled without problems.
Your compiled G5WS work fine, but i`d like to experiment

**Reply by PhlexPlexico** [reply]
This is due to some functionality being changed in get5. Version 4.0.0 will only work with get5 0.10 and higher. Please look at what the workflow script does to compile this plugin.

https://github.com/PhlexPlexico/G5WS/blob/steamworks-reimplement/.github/workflows/make_release.yml

That should help you compile it as needed.

**Reply by PhlexPlexico** [reply]
Also it looks like some more code was moved from the utils calls. The latest push should fix this  issue.

**Reply by Bunker56** [reply]
Compiled, trying...
